### DISCUSSION TITLE
Omitting the `Attribute` suffix when using an attribute class as a type argument

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#5675] [Language Ideas]
Generic symbols (methods or types) that contain a generic type parameter constrained to being an attribute type do not allow simplifying the attribute class name. However, only attribute classes can be used as type arguments for those type parameters. Therefore, there is enough context for the binder to know whether we're dealing with a shortened attribute class name or not.

For example, `System.Reflection` contains some extension methods like `GetCustomAttribute<T>() where T : Attribute`. It would be handy to be able to write `typeof(T).GetCustomAttribute<Serializable>()`, instead of needing to include the `Attribute` suffix on `Serializable`.

## Drawbacks

- This may affect maintainability in certain contexts where the generic type parameter's constraints are changed. Code is more prone to breaking changes from such actions, in cases where the shortened version of the attribute class' name is used for another type.
- There aren't too many type parameters constrained to `Attribute`. Such a feature would be too niche in the real world.

**Comment by huoyaoyuan** [comment]
Type argument is too large a scope, comparing to using attribute in attribute syntax. Adding special cases will make things worse.

**Comment by HaloFour** [comment]
The reason you can omit the `Attribute` suffix of an attribute class when it is applies isn't to make it shorter, it's to make it an adjective.  That consideration doesn't apply when you're working with the attribute as a type.  It's also why the ability to omit a suffix doesn't exist anywhere else in the language.
